article/repository.go

```diff
@@ -14,4 +14,5 @@ type Repository interface {
 	Update(ctx context.Context, ar *models.Article) error
 	Store(ctx context.Context, a *models.Article) error
 	Delete(ctx context.Context, id int64) error
+	WithTransaction(func(Repository)error) error
```

Comment:
I wonder, how we use this in usecase/service layer?

Reply:
```go
err = s.repo.WithTransaction( func(repo aritcle.Repository) error  {
	if err := repo.Create(c, article); err != nil {
		return err
	}
        return nil
})
```
use it like this, its more clear

Reply:
Yeah... I  see this more simple. 
But how do you handle like transaction across domain. 

Right now, it's okay since it's still the same article (in one repository) 
How about across to another repository, like to author repository, or maybe publisher repository, category repository?


Let's say, I want to store article, including the category, the author, then the publisher?

Reply:
I think one transaction should only change one aggregate root, and we should have only one repo for one aggregate root.
if your operation across other aggregate root, try eventual consistency instead of strong consistency

Reply:
By saying that, so will you use a pub/sub pattern for that from the app level. Or from infra levels like using 3rd party message broker, and pubsub?

Reply:
A message broker is better option in micro service architecture

Reply:
You can split `WithTransaction` form the Repository as an utility. Take a look at `transactor` example https://github.com/r3code/go-useful-snippets/blob/master/dbutils/transactor/transaction_integration_test.go

Reply:
@r3code
This is new. This will decouple everything. Thanks for the idea. 

What do you think @philchia?
*anyway, I'll keep this PR open. Since there's a lot of options to do the transactions here. If someone in the future has another idea, just keep posting on this PR. Just let's make this as a knowledge base.

Reply:
How would you use the same transaction across multiple repositories? 

The way the `okFunc` and `panicFunc` works in the linked test is that it runs the repository func with the transaction then commits or rollbacks based on a panic after only one query.

Reply:
I also got some inspiration from https://medium.com/@jfeng45/go-microservice-with-clean-architecture-transaction-support-61eb0f886a36, 

I really liked it where it states that transaction logic should be applied to the use case layer, and not the persistence layer.

Basically, what he did is to create another wrapper around sql.DB and sql.Tx

https://gist.github.com/adamfdl/49da2b850991c59939da2426fa85c1ab

Reply:
Thanks for the idea @adamfdl 

I'm thinking of a general-purpose of architecture here. Luckily you use SQL which RDBMS here that supports the transaction features so that we can wrap the sql.DB and sql.Tx 

But for RDBMS, I think this a good solution.

Reply:
@adamfdl can you post a more complete example, please? How to use `Wrap()` in a usecase for example? Thanks, really.